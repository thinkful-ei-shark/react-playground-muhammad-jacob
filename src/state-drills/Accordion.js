import React, { useState } from 'react';


//Function based
export default function Accordion(props) {
  const [buttonClickedIndex, setButtonClickedIndex] = useState(null);

  let toggleButton = (e={currentTarget:{getAttribute:()=>0}}) => {
    buttonClickedIndex !== null && buttonClickedIndex === Number(e.currentTarget.getAttribute('index'))
      ? setButtonClickedIndex(null)
      : setButtonClickedIndex(Number(e.currentTarget.getAttribute('index')))
  }

  let renderList = () => {
    return props.sections.map((section, index) => {
      return (
        <li key={index}>
          {renderButtons(section, index)}
          {buttonClickedIndex === index ? getContent() : void 0}
        </li>
      )
    })
  }

  let renderButtons = (section, index) => {
    return (
      <button key={index} index={index} onClick={toggleButton}>{section.title}</button>
    )
  }

  let getContent = () => {
    return (
      <p>
        {props.sections[buttonClickedIndex].content}
      </p>
    )
  }

  return (
    <ul>{renderList()}</ul>
  )
}

Accordion.defaultProps = { sections: [] };

//Class based

/* export default class Accordion extends React.Component {

  static defaultProps = { sections: [] }

  state = {
    buttonClickedIndex: null
  }

  toggleButton = e => {
    this.state.buttonClickedIndex !== null
      ? this.setState({ buttonClickedIndex: null })
      : this.setState({ buttonClickedIndex: Number(e.currentTarget.getAttribute('index')) })
  }

  renderList = () => {
    return this.props.sections.map((section, index) => {
      return (
        <li>
          {this.renderButtons(section, index)}
          {this.state.buttonClickedIndex === index ? this.getContent() : void 0}
        </li>
      )
    })
  }

  renderButtons = (section, index) => {
    return (
      <button key={index} index={index} onClick={this.toggleButton}>{section.title}</button>
    )
  }

  getContent = () => {
    return (
      <p>
        {this.props.sections[this.state.buttonClickedIndex].content}
      </p>
    )
  }

  render() {
    return (
      <ul>{this.renderList()}</ul>
    )
  }

} */