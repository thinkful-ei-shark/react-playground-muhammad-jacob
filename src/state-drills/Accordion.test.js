import React from 'react';
import ReactDOM from 'react-dom';
import Accordion from './Accordion';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const sections = [
  {
    title: 'Section 1',
    content: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
  },
  {
    title: 'Section 2',
    content: 'Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
  },
  {
    title: 'Section 3',
    content: 'Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
  },
]

describe('Accordion Component', () => {
  it('should render to the UI', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Accordion />, div);
    ReactDOM.unmountComponentAtNode(div);
  })

  it('should render empty with no supplied sections prop', () => {
    const json = toJson(shallow(<Accordion />));
    expect(json).toMatchSnapshot();
  })

  it('should render just the buttons, none expanded', () => {
    const json = toJson(shallow(<Accordion sections={sections} />));
    expect(json).toMatchSnapshot();
  })

  it('should render first button expanded', () => {
    const wrapper = shallow(<Accordion sections={sections} />);
    wrapper.find('button').at(1).simulate('click');
    const json = toJson(wrapper);
    expect(json).toMatchSnapshot();
  })

  it('should render expanded of the last button that was clicked', () => {
    const wrapper = shallow(<Accordion sections={sections} />);
    wrapper.find('button').at(1).simulate('click');
    wrapper.find('button').at(2).simulate('click');
    const json = toJson(wrapper);
    expect(json).toMatchSnapshot();
  })

})