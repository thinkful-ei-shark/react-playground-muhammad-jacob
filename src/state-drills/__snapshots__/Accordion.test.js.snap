// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component should render empty with no supplied sections prop 1`] = `<ul />`;

exports[`Accordion Component should render expanded of the last button that was clicked 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      index={0}
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      index={1}
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      index={2}
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component should render first button expanded 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      index={0}
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
    <p>
      Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      index={1}
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      index={2}
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component should render just the buttons, none expanded 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      index={0}
      key="0"
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      index={1}
      key="1"
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      index={2}
      key="2"
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;
